addons/project/models/project_task_recurrence.py

```diff
@@ -63,30 +63,49 @@ def _get_recurrence_delta(self):
             f"{self.repeat_unit}s": self.repeat_interval
         })
 
-    def _create_next_occurrence(self, occurrence_from):
-        self.ensure_one()
-        if self.repeat_type == 'until' and fields.Date.today() > self.repeat_until:
-            return
-        occurrence_from.with_context(copy_project=True).sudo().copy(self._create_next_occurrence_values(occurrence_from))
+    @api.model
+    def _create_next_occurrences(self, occurrences_from):
+        occurrences_from = occurrences_from.filtered(lambda task:
+            task.recurrence_id.repeat_type != 'until' or
+            task.recurrence_id.repeat_until >= fields.Date.today()
+        )
+        if occurrences_from:
+            task_by_recurrence = {task: task.recurrence_id for task in occurrences_from}
```

Comment:
nitpick, but i'd call this dict 'recurrence by task' instead, since the task are the keys and the recurrence the values.